File: mobile/app/src/main/java/com/remotedesktop/webrtc/WebRTCClient.kt
package com.remotedesktop.webrtc

import android.content.Context
import com.remotedesktop.signaling.IceCandidatePayload
import com.remotedesktop.signaling.SdpPayload
import org.webrtc.AudioTrack
import org.webrtc.DataChannel
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.SdpObserver
import org.webrtc.SessionDescription
import org.webrtc.VideoTrack
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import java.util.concurrent.atomic.AtomicBoolean

class WebRTCClient(
    context: Context,
    private val eglBase: EglBase,
    private val callbacks: Callbacks
) {
    interface Callbacks {
        fun onLocalIceCandidate(candidate: IceCandidatePayload)
        fun onRemoteVideoTrack(track: VideoTrack)
        fun onDataChannel(channel: DataChannel)
        fun onConnectionStateChange(state: PeerConnection.PeerConnectionState)
    }

    private val factory: PeerConnectionFactory
    private var peerConnection: PeerConnection? = null
    private val closed = AtomicBoolean(false)

    init {
        PeerConnectionFactory.initialize(
            PeerConnectionFactory.InitializationOptions.builder(context)
                .createInitializationOptions()
        )
        val encoder = DefaultVideoEncoderFactory(eglBase.eglBaseContext, true, true)
        val decoder = DefaultVideoDecoderFactory(eglBase.eglBaseContext)
        factory = PeerConnectionFactory.builder()
            .setVideoEncoderFactory(encoder)
            .setVideoDecoderFactory(decoder)
            .createPeerConnectionFactory()
    }

    fun createConnection(iceServers: List<PeerConnection.IceServer>) {
        val rtcConfig = PeerConnection.RTCConfiguration(iceServers).apply {
            sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
            tcpCandidatePolicy = PeerConnection.TcpCandidatePolicy.DISABLED
            iceTransportsType = PeerConnection.IceTransportsType.ALL
        }
        peerConnection = factory.createPeerConnection(rtcConfig, object : PeerConnection.Observer {
            override fun onSignalingChange(state: PeerConnection.SignalingState) {}
            override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {}
            override fun onIceConnectionReceivingChange(receiving: Boolean) {}
            override fun onIceGatheringChange(state: PeerConnection.IceGatheringState) {}
            override fun onIceCandidate(candidate: IceCandidate) {
                callbacks.onLocalIceCandidate(IceCandidatePayload(
                    candidate.sdp, candidate.sdpMid, candidate.sdpMLineIndex
                ))
            }
            override fun onIceCandidatesRemoved(candidates: Array<out IceCandidate>?) {}
            override fun onAddStream(stream: MediaStream?) {}
            override fun onRemoveStream(stream: MediaStream?) {}
            override fun onDataChannel(channel: DataChannel) {
                callbacks.onDataChannel(channel)
            }
            override fun onRenegotiationNeeded() {}
            override fun onAddTrack(receiver: RtpReceiver?, streams: Array<out MediaStream>?) {
                val track = receiver?.track() ?: return
                when (track) {
                    is VideoTrack -> callbacks.onRemoteVideoTrack(track)
                    is AudioTrack -> {}
                }
            }
            override fun onConnectionChange(state: PeerConnection.PeerConnectionState) {
                callbacks.onConnectionStateChange(state)
            }
        }) ?: error("Failed to create peer connection")
    }

    fun setRemoteOffer(payload: SdpPayload, onSuccess: () -> Unit, onError: (String) -> Unit) {
        val description = SessionDescription(SessionDescription.Type.fromCanonicalForm(payload.type), payload.sdp)
        val pc = peerConnection ?: return onError("peer connection not initialised")
        pc.setRemoteDescription(simpleObserver(onSuccess, onError), description)
    }

    fun createAnswer(onSuccess: (SdpPayload) -> Unit, onError: (String) -> Unit) {
        val pc = peerConnection ?: return onError("peer connection not initialised")
        val constraints = MediaConstraints()
        pc.createAnswer(object : SdpObserver {
            override fun onCreateSuccess(description: SessionDescription) {
                pc.setLocalDescription(simpleObserver({
                    onSuccess(SdpPayload(description.type.canonicalForm(), description.description))
                }, onError), description)
            }
            override fun onSetSuccess() {}
            override fun onCreateFailure(error: String) { onError(error) }
            override fun onSetFailure(error: String) { onError(error) }
        }, constraints)
    }

    fun addRemoteIceCandidate(payload: IceCandidatePayload) {
        val pc = peerConnection ?: return
        val candidate = IceCandidate(payload.sdpMid ?: "", payload.sdpMLineIndex ?: 0, payload.candidate)
        pc.addIceCandidate(candidate)
    }

    fun sendData(channel: DataChannel, message: String) {
        val bytes = message.toByteArray(StandardCharsets.UTF_8)
        val buffer = DataChannel.Buffer(ByteBuffer.wrap(bytes), false)
        channel.send(buffer)
    }

    fun dispose() {
        if (!closed.compareAndSet(false, true)) return
        try { peerConnection?.close() } catch (_: Exception) {}
        peerConnection = null
        try { factory.dispose() } catch (_: Exception) {}
    }

    private fun simpleObserver(onSuccess: () -> Unit, onError: (String) -> Unit): SdpObserver =
        object : SdpObserver {
            override fun onCreateSuccess(description: SessionDescription?) {}
            override fun onSetSuccess() { onSuccess() }
            override fun onCreateFailure(error: String) { onError(error) }
            override fun onSetFailure(error: String) { onError(error) }
        }

    companion object {
        val DEFAULT_ICE_SERVERS: List<PeerConnection.IceServer> = listOf(
            PeerConnection.IceServer.builder("stun:stun.l.google.com:19302").createIceServer(),
            PeerConnection.IceServer.builder("stun:stun1.l.google.com:19302").createIceServer()
        )
    }
}

File: mobile/ios/RemoteDesktop/Sources/WebRTCClient.swift
import Foundation
import WebRTC

protocol WebRTCClientDelegate: AnyObject {
    func webRTC(_ client: WebRTCClient, didDiscoverLocalCandidate candidate: IceCandidatePayload)
    func webRTC(_ client: WebRTCClient, didReceiveRemoteTrack track: RTCVideoTrack)
    func webRTC(_ client: WebRTCClient, didOpenDataChannel channel: RTCDataChannel)
    func webRTC(_ client: WebRTCClient, didChangeConnectionState state: RTCPeerConnectionState)
    func webRTC(_ client: WebRTCClient, didReceiveMessage message: String)
}

final class WebRTCClient: NSObject, RTCPeerConnectionDelegate, RTCDataChannelDelegate {
    weak var delegate: WebRTCClientDelegate?

    static let defaultIceServers = [
        RTCIceServer(urlStrings: ["stun:stun.l.google.com:19302", "stun:stun1.l.google.com:19302"])
    ]

    private static let factory: RTCPeerConnectionFactory = {
        RTCInitializeSSL()
        let encoder = RTCDefaultVideoEncoderFactory()
        let decoder = RTCDefaultVideoDecoderFactory()
        return RTCPeerConnectionFactory(encoderFactory: encoder, decoderFactory: decoder)
    }()

    private var connection: RTCPeerConnection?
    private var inputChannel: RTCDataChannel?

    func createConnection(iceServers: [RTCIceServer] = defaultIceServers) {
        let config = RTCConfiguration()
        config.iceServers = iceServers
        config.sdpSemantics = .unifiedPlan
        config.tcpCandidatePolicy = .disabled

        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        connection = Self.factory.peerConnection(with: config, constraints: constraints, delegate: self)
    }

    func setRemoteOffer(_ payload: SdpPayload, completion: @escaping (Error?) -> Void) {
        let description = RTCSessionDescription(type: .offer, sdp: payload.sdp)
        connection?.setRemoteDescription(description, completionHandler: completion)
    }

    func createAnswer(completion: @escaping (Result<SdpPayload, Error>) -> Void) {
        guard let pc = connection else {
            completion(.failure(NSError(domain: "WebRTC", code: -1)))
            return
        }
        let constraints = RTCMediaConstraints(mandatoryConstraints: nil, optionalConstraints: nil)
        pc.answer(for: constraints) { description, error in
            if let error = error { return completion(.failure(error)) }
            guard let description = description else {
                return completion(.failure(NSError(domain: "WebRTC", code: -2)))
            }
            pc.setLocalDescription(description) { error in
                if let error = error { return completion(.failure(error)) }
                completion(.success(SdpPayload(type: "answer", sdp: description.sdp)))
            }
        }
    }

    func addRemoteIceCandidate(_ payload: IceCandidatePayload) {
        let candidate = RTCIceCandidate(
            sdp: payload.candidate,
            sdpMLineIndex: Int32(payload.sdpMLineIndex ?? 0),
            sdpMid: payload.sdpMid
        )
        connection?.add(candidate, completionHandler: { _ in })
    }

    func sendData(_ message: String) {
        guard let channel = inputChannel, channel.readyState == .open else { return }
        let buffer = RTCDataBuffer(data: Data(message.utf8), isBinary: false)
        channel.sendData(buffer)
    }

    func dispose() {
        inputChannel?.close()
        inputChannel = nil
        connection?.close()
        connection = nil
    }

    // MARK: - RTCPeerConnectionDelegate

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCSignalingState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd stream: RTCMediaStream) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove stream: RTCMediaStream) {}
    func peerConnectionShouldNegotiate(_ peerConnection: RTCPeerConnection) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceConnectionState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCIceGatheringState) {}
    func peerConnection(_ peerConnection: RTCPeerConnection, didRemove candidates: [RTCIceCandidate]) {}

    func peerConnection(_ peerConnection: RTCPeerConnection, didGenerate candidate: RTCIceCandidate) {
        delegate?.webRTC(self, didDiscoverLocalCandidate: IceCandidatePayload(
            candidate: candidate.sdp,
            sdpMid: candidate.sdpMid,
            sdpMLineIndex: Int(candidate.sdpMLineIndex)
        ))
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didOpen dataChannel: RTCDataChannel) {
        inputChannel = dataChannel
        dataChannel.delegate = self
        delegate?.webRTC(self, didOpenDataChannel: dataChannel)
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didAdd rtpReceiver: RTCRtpReceiver, streams mediaStreams: [RTCMediaStream]) {
        if let track = rtpReceiver.track as? RTCVideoTrack {
            delegate?.webRTC(self, didReceiveRemoteTrack: track)
        }
    }

    func peerConnection(_ peerConnection: RTCPeerConnection, didChange newState: RTCPeerConnectionState) {
        delegate?.webRTC(self, didChangeConnectionState: newState)
    }

    // MARK: - RTCDataChannelDelegate

    func dataChannelDidChangeState(_ dataChannel: RTCDataChannel) {}

    func dataChannel(_ dataChannel: RTCDataChannel, didReceiveMessageWith buffer: RTCDataBuffer) {
        let message = String(data: buffer.data, encoding: .utf8) ?? ""
        delegate?.webRTC(self, didReceiveMessage: message)
    }
}
